core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncPathCache.java

```diff
@@ -65,11 +78,62 @@ public boolean shouldSyncPath(String path, long intervalMs) {
       // Always sync.
       return true;
     }
-    Long lastSync = mCache.getIfPresent(path);
-    if (lastSync == null) {
-      // No info about the last sync, so trigger a sync.
-      return true;
+
+    // check the last sync information for the path itself.
+    SyncTime lastSync = mCache.getIfPresent(path);
+    if (lastSync != null && (System.currentTimeMillis() - lastSync.getLastSyncMs()) < intervalMs) {
+      // Sync is not necessary for this path.
+      return false;
+    }
+
+    // sync should be done on this path, but check all ancestors to determine if a recursive sync
+    // had been performed (to avoid a sync again).
+    String currPath = path;
+    while (!currPath.equals(AlluxioURI.SEPARATOR)) {
```

Comment:
perhaps use `isRoot`? also i think we also need to check if root was actually synced as well.

Reply:
Well, `path` comes in as a `String` and I didn't want to keep converting to and from `AlluxioURI`. I think that could be too expensive. Also, root will be checked, because within the loop, the parent of the path is checked. So, the root will be checked on the previous iteration.

Reply:
oh that makes sense